java/query/GraqlDefinable.java

```diff
@@ -30,43 +33,52 @@
 import static graql.lang.common.GraqlToken.Char.SEMICOLON;
 import static graql.lang.common.GraqlToken.Command.DEFINE;
 import static graql.lang.common.GraqlToken.Command.UNDEFINE;
-import static graql.lang.common.exception.ErrorMessage.MISSING_PATTERNS;
+import static graql.lang.common.exception.ErrorMessage.MISSING_DEFINABLES;
 import static java.util.stream.Collectors.joining;
 
 abstract class GraqlDefinable extends GraqlQuery {
 
     private final GraqlToken.Command keyword;
-    private final List<TypeVariable> variables;
+    private final List<TypeVariable> typeVariables = new ArrayList<>();
+    private final List<Rule> rules = new ArrayList<>();
     private final int hash;
 
-    GraqlDefinable(final GraqlToken.Command keyword, final List<TypeVariable> variables) {
+    GraqlDefinable(final GraqlToken.Command keyword, final List<Definable> definables) {
         assert keyword == DEFINE || keyword == UNDEFINE;
-        if (variables == null || variables.isEmpty()) throw GraqlException.of(MISSING_PATTERNS.message());
-        final LinkedList<TypeVariable> list = new LinkedList<>(variables);
-        while (!list.isEmpty()) {
-            final TypeVariable v = list.removeFirst();
+        if (definables == null || definables.isEmpty()) throw GraqlException.of(MISSING_DEFINABLES.message());
+        final LinkedList<TypeVariable> typeVars = new LinkedList<>();
+        for (Definable definable : definables) {
+            if (definable.isRule()) rules.add(definable.asRule());
+            if (definable.isTypeVariable()) typeVars.add(definable.asTypeVariable());
+        }
+        while (!typeVars.isEmpty()) {
+            final TypeVariable v = typeVars.removeFirst();
             if (!v.isLabelled()) throw GraqlException.of(ErrorMessage.INVALID_DEFINE_QUERY_VARIABLE.message());
-            else v.constraints().forEach(c -> list.addAll(c.variables()));
+            else v.constraints().forEach(c -> typeVars.addAll(c.variables()));
+            typeVariables.add(v);
         }
 
         this.keyword = keyword;
-        this.variables = variables;
-        this.hash = Objects.hash(this.keyword, this.variables);
+        this.hash = Objects.hash(this.keyword, this.typeVariables, this.rules);
     }
 
-    public final List<TypeVariable> variables() {
-        return variables;
+    public final List<TypeVariable> typeVariables() {
+        return typeVariables;
+    }
```

Comment:
I think this method can simply be called `variables()` as the context is clear: it is in `GraqlDefinable`. The only type of variable that can exist here is only `TypeVariable`. Same goes for `GraqlWritable` - only `ThingVariable` exists there, so that's what their `variables()` will return. We can revert the name of the class field too.

Reply:
was debating those choices with myself too! ok